### PR TITLE
feat(launch): Sweeps on Launch MVP

### PR DESCRIPTION
**Pull request by hu-po**
WB-9692
WB-9592 (EPIC)

Description
-----------

[Notion Doc](https://www.notion.so/wandbai/Sweeps-on-Launch-e64c921b0c9f4b9d903ba6d136885a0d)

This PR introduces a sweep+launch `Scheduler`, which populates a launch job queue with sweep jobs that can be read and consumed by a Launch Agent. We created a sweep with the following command (note the new `--queue` arg):

`wandb sweep sweep-bayes.yaml --queue default`

This puts a `Scheduler` job on the default launch queue. We can then run a launch agent to start consuming work from that queue:

`wandb launch-agent -q default -p sweeps-examples`

The launch agent will run the `Scheduler`, which will use `AgentHeartbeat` to get run suggestions via the normal Anaconda2 sweeps codepath. These run suggestions will be packaged by the `Scheduler` into launch jobs, which will then be put into the `default` launch job queue. Launch jobs are run using the `LocalProcessRunner` backend.

Testing
-------
```
# Included tests
pytest -vs -- tests/tests_launch/test_launch_sweeps.py

# Bayes search locally on default queue (Happy path usecase)
WANDB_BASE_URL=https://api.wandb.test wandb sweep sweep-bayes.yaml --queue default --entity hupo
WANDB_BASE_URL=https://api.wandb.test wandb launch-agent -q my_cool_sweep -p sweeps-examples

# Bayes search locally on user-created queue
WANDB_BASE_URL=https://api.wandb.test wandb sweep sweep-bayes.yaml --queue my_cool_queue --entity hupo
WANDB_BASE_URL=https://api.wandb.test wandb launch-agent -q my_cool_queue -p sweeps-examples

# Grid search locally on default queue (Parallelism usecase)
WANDB_BASE_URL=https://api.wandb.test wandb sweep sweep-grid.yaml --queue default --entity hupo
WANDB_BASE_URL=https://api.wandb.test wandb launch-agent -q default -p sweeps-examples -j -1

```

### CODE REVIEW

File: wandb/cli/cli.py [hunk 1]
```diff
@@ -1144,7 +1214,13 @@ def launch(
 
 @cli.command(context_settings=CONTEXT, help="Run a W&B launch agent (Experimental)")
 @click.pass_context
-@click.argument("project", nargs=1, required=False)
+@click.option(
+    "--project",
+    "-p",
+    default=None,
+    help="Name of the target project which the new run will be sent to. "
+    "If passed in, will override the project value passed in using a config file.",
```
Comment (KyleGoyette) [inline wandb/cli/cli.py:1222]: this is the project that the agent will watch rather than the project the run will be sent to. (They are one in the same right now but won't alwasy be)

File: wandb/cli/cli.py [hunk 2]
```diff
@@ -1185,9 +1264,20 @@ def launch_agent(
...
+        # the launch scheduler will become a type of Job (Launch Jobs API)
+        # until then this felt like the least worst way to start launch schedulers.
+        wandb.termlog("Starting a Launch Scheduler 🚀")
+        from wandb.sdk.launch.sweeps import load_scheduler
 
-    wandb_launch.create_and_run_agent(api, agent_config)
+        _scheduler = load_scheduler("sweep")(
+            api, entity=entity, project=project, queue=queues[0], sweep_id=sweep_id
+        )
+        _scheduler.start()
```
Comment (KyleGoyette) [inline wandb/cli/cli.py:1277]: if this is being started in a second process, will the launch agent have a way to kill it on KeyboardInterupt?

File: wandb/sdk/internal/internal_api.py [hunk 3]
```diff
@@ -956,7 +956,10 @@ def push_to_run_queue(self, queue_name, launch_spec):
             for q in queues_found
             if q["name"] == queue_name
             # ensure user has access to queue
-            and (q["access"] == "PROJECT" or q["createdBy"] == self.default_entity)
+            and (
+                q["access"] in ["PROJECT", "USER"]
```
Comment (KyleGoyette) [inline wandb/sdk/internal/internal_api.py:960]: Why this change? This will give access to all queues, since queue access is either PROJECT or USER

File: wandb/sdk/launch/agent/agent.py [hunk 4]
```diff
@@ -144,9 +145,9 @@ def _validate_and_fix_spec_project_entity(
 
     def run_job(self, job: Dict[str, Any]) -> None:
         """Sets up project and runs the job."""
-        # TODO: logger
-        wandb.termlog(f"agent: got job f{job}")
-        _logger.info(f"Agent job: {job}")
+        _msg = f"Launch agent received job:\n{pprint.pformat(job)}\n"
```
Comment (KyleGoyette) [inline wandb/sdk/launch/agent/agent.py:148]: 🔥

File: wandb/sdk/launch/launch.py [hunk 5]
```diff
@@ -26,7 +26,7 @@ def resolve_agent_config(
     api: Api,
     entity: Optional[str],
     project: Optional[str],
-    max_jobs: Optional[float],
+    max_jobs: Optional[int],
```
Comment (KyleGoyette) [inline wandb/sdk/launch/launch.py:29]: Hmm, you know I thought we set this to float because -1 was infinite jobs and -1 was a float... but checking it its an int. Nice fix, but can you make sure -1 works for max_jobs?

File: wandb/sdk/launch/runner/local_process.py [hunk 6]
```diff
@@ -50,15 +51,18 @@ def run(  # type: ignore
...
-        # Check to make sure local python dependencies match run's requirement.txt
-        source_entity, source_project, run_name = parse_wandb_uri(launch_project.uri)
-        validate_wandb_python_deps(
-            source_entity,
-            source_project,
-            run_name,
-            self._api,
-            launch_project.project_dir,
-        )
+        # If URI is not a wandb run, heck to make sure local python dependencies match run's requirement.txt
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/local_process.py:54]: check*

File: wandb/sdk/launch/sweeps/scheduler.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+            return False
+        return True
+
+    def start(self) -> None:
+        _msg = "Scheduler starting."
+        logger.debug(_msg)
+        wandb.termlog(_msg)
+        self._state = SchedulerState.STARTING
+        self._start()
+        # TODO(hupo): Should start call run?
```
Comment (KyleGoyette) [inline wandb/sdk/launch/sweeps/scheduler.py:115]: Yeah? where else would you call run?

File: wandb/sdk/launch/_project_spec.py [hunk 8]
```diff
@@ -84,7 +85,7 @@ def __init__(
         self.deps_type: Optional[str] = None
         self.cuda = cuda
         self._runtime: Optional[str] = None
-        self.run_id = generate_id()
+        self.run_id = run_id or generate_id()
```
[the remark this answers is not in the document]
Comment (hu-po) [reply wandb/sdk/launch/_project_spec.py:88]: Added TODO

File: wandb/sdk/internal/internal_api.py (hunk 3 above)
Comment (hu-po) [reply wandb/sdk/internal/internal_api.py:960]: If I don't set this I get "queue not found" errors with a launch queue I create through the UI. Guessing that shouldn't be the case?

File: wandb/sdk/launch/sweeps/scheduler.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        run_id = run_id or generate_id()
+        job = launch_add(
+            uri,
+            project=self._project,
+            entity=self._entity,
+            queue=self._launch_queue,
+            resource=resource,
+            entry_point=entry_point,
+            run_id=run_id,
+            # config: Optional[Union[str, Dict[str, Any]]] = None,
```
Comment (KyleGoyette) [inline wandb/sdk/launch/sweeps/scheduler.py:198]: nit: I think these comments should be removed

**Review by KyleGoyette — COMMENTED** [review]
I think this generally looks all right. I am concerned about the URI being the local directory. Since the jobs will be run at the location of the launch agent which could be run anywhere. Jobs will offer a potential solution to this issue, I think at this time we'll need to warn users to only run the launch agent at the same location as the sweep.

File: wandb/cli/cli.py [hunk 10]
```diff
@@ -856,8 +869,58 @@ def _parse_settings(settings):
...
+                        None,  # resource_args,
+                        None,  # launch_config,
+                        None,  # cuda,
+                        None,  # run_id,
+                    )
+                ),
+            }
+        )
+
     sweep_id, warnings = api.upsert_sweep(
```
Comment (jlzhao27) [inline wandb/cli/cli.py:918]: how do we know what command to run to start the actual sweep job here?

File: wandb/cli/cli.py [hunk 11]
```diff
@@ -1185,9 +1273,20 @@ def launch_agent(
             "You must specify a project name or set WANDB_PROJECT environment variable."
         )
 
-    wandb.termlog("Starting launch agent ✨")
+    if sweep_id is not None:
```
Comment (jlzhao27) [inline wandb/cli/cli.py:1276]: hmm i think if this is a short term hack so we have a valid entrypoint into the sweep agent, we should move it to the `wandb sweep-agent` command or make a new command. I don't want to pollute the launch-agent command with args that we may remove in the future since this is the code path we want to keep clean

File: wandb/sdk/internal/internal_api.py (hunk 3 above)
Comment (jlzhao27) [reply wandb/sdk/internal/internal_api.py:979]: sounds like this should be reverted then?

File: wandb/cli/cli.py [hunk 12]
```diff
@@ -856,8 +869,58 @@ def _parse_settings(settings):
...
+                    construct_launch_spec(
+                        os.getcwd(),  # uri,
+                        api,
+                        f"Scheduler.{_sweep_id_placeholder}",  # name,
+                        project,
+                        entity,
+                        None,  # docker_image,
+                        "local-process",  # resource,
+                        [
+                            "wandb",
```
Comment (jlzhao27) [inline wandb/cli/cli.py:898]: hmm, I see that you hijacked the `wandb launch-agent` comment with a `--queues` arg which essentially starts the sweep scheduler. Similar to my comment below, I would prefer moving this to another sub command in the SDK, probably the `wandb sweep-agent` would be my preference so we can keep `wandb launch-agent` clean

File: wandb/sdk/launch/sweeps/scheduler.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+import wandb
+from wandb.apis.internal import Api
+import wandb.apis.public as public
+from wandb.sdk.launch.launch_add import launch_add
+from wandb.sdk.lib.runid import generate_id
+
+logger = logging.getLogger(__name__)
+
+
+class SchedulerState(Enum):
```
Comment (jlzhao27) [inline wandb/sdk/launch/sweeps/scheduler.py:17]: is this needed or can this be reflected via the launch run states? since the scheduler itself is just a run

File: wandb/sdk/launch/sweeps/scheduler.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        wandb.termlog(_msg)
+        return job
+
+    def _stop_run(self, run_id: str) -> None:
+        _msg = f"Stopping run {run_id}."
+        logger.debug(_msg)
+        wandb.termlog(_msg)
+        run = self._runs.get(run_id, None)
+        if run is not None:
+            run.state = SimpleRunState.DEAD
```
Comment (jlzhao27) [inline wandb/sdk/launch/sweeps/scheduler.py:206]: is this actually how you stop a run lol, don't you need to make an API call to launch to kill it?

File: wandb/sdk/launch/sweeps/scheduler.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+class SweepRun:
+    id: str
+    state: SimpleRunState = SimpleRunState.ALIVE
+    launch_job: Optional[public.QueuedJob] = None
+    args: Optional[Dict[str, Any]] = None
+    logs: Optional[List[str]] = None
+    program: Optional[str] = None
+
+
+class Scheduler(ABC):
```
Comment (jlzhao27) [inline wandb/sdk/launch/sweeps/scheduler.py:42]: very cool class! I like how it is becoming much more general :)

**Review by jlzhao27 — COMMENTED** [review]
looks pretty good overall, I didn't review `scheduler_sweep.py` at all since that is all localized sweep code. The main feedback I have is we should change the command to start a sweep scheduler, I just don't want us to pollute the `wandb launch-agent` command (And agents running other agents is conceptually very confusing)

File: wandb/sdk/launch/sweeps/scheduler_sweep.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            _msg = f"AgentHeartbeat sending: \n{pprint.pformat(_run_states)}\n"
+            logger.debug(_msg)
+            wandb.termlog(_msg)  # TODO: Remove
+            commands = self._api.agent_heartbeat(
+                self._heartbeat_agent_id, {}, _run_states
+            )
+            if commands:
+                _msg = f"AgentHeartbeat received {len(commands)} commands: \n{pprint.pformat(commands)}\n"
+                logger.debug(_msg)
+                wandb.termlog(_msg)  # TODO: Remove
```
Comment (KyleGoyette) [inline wandb/sdk/launch/sweeps/scheduler_sweep.py:82]: do not forget to remove

File: wandb/sdk/launch/sweeps/scheduler_sweep.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        # This is actually what populates the wandb config
+        # since it is used in wandb.init()
+        sweep_param_path = os.path.join(
+            os.environ.get(wandb.env.DIR, os.getcwd()),
+            "wandb",
+            f"sweep-{self._sweep_id}",
+            f"config-{run.id}.yaml",
+        )
+        wandb.termlog(f"Saving params to {sweep_param_path}")
+        wandb_lib.config_util.save_config_file_from_dict(sweep_param_path, run.args)
```
Comment (KyleGoyette) [inline wandb/sdk/launch/sweeps/scheduler_sweep.py:133]: We will eventually need to move the sweep config stuff to an environment variable. Since we're running with launch, this should just work right now, but will need to be done when we move away from local_process

Comment (KyleGoyette) [reply wandb/sdk/launch/sweeps/scheduler_sweep.py:133]: All you should need to do is put the config here into the override configs in the add to launch queue call below

File: wandb/cli/cli.py (hunk 11 above)
Comment (hu-po) [reply wandb/cli/cli.py]: See above.

File: wandb/sdk/launch/sweeps/scheduler.py (hunk 13 above)
Comment (hu-po) [reply wandb/sdk/launch/sweeps/scheduler.py:17]: Yeah there is definitely a bunch of overlap between states for runs, jobs, scheduler, etc. I figured having a separate state for the Scheduler would be useful in the future when there are a bunch of types of schedulers. But we can always change it to match the Job state in the future.

File: wandb/sdk/launch/sweeps/scheduler_sweep.py (hunk 17 above)
Comment (hu-po) [reply wandb/sdk/launch/sweeps/scheduler_sweep.py:133]: Okay, I piped it through `params` which seems to get merged into override configs

File: wandb/sdk/launch/sweeps/scheduler.py (hunk 14 above)
Comment (hu-po) [reply wandb/sdk/launch/sweeps/scheduler.py:206]: The scheduler can only put the runs on the queue and then query their state via the backend. I think only the launch-agent has the handle to the actual processes to kill them: https://github.com/wandb/client/blob/225c28f6f62519b6acc2a93aa4fcc20d0ffec748/wandb/sdk/launch/agent/agent.py#L235.

File: wandb/sdk/internal/internal_api.py [hunk 18]
```diff
@@ -1901,6 +1907,7 @@ def upsert_sweep(
                 entityName: $entityName,
                 projectName: $projectName,
                 controller: $controller,
+                launchScheduler: $launchScheduler,
```
Comment (raubitsj) [inline wandb/sdk/internal/internal_api.py:1910]: I am pretty sure this will break every `wandb server` user who has a build before:
https://github.com/wandb/core/commit/13a1c318816182c25420a6c8f3ef00c2a7e3b508
... which is all of them.

You will want to gate additions to graphql mutations and queries by introspection.

**Review by raubitsj — CHANGES_REQUESTED** [review]

File: wandb/sdk/internal/internal_api.py (hunk 18 above)
Comment (hu-po) [reply wandb/sdk/internal/internal_api.py]: Okay fixed it by adding another mutation to the `for mutation in mutations` logic that exists here in the UpsertSweep mutation.

File: wandb/sdk/internal/internal_api.py [hunk 19]
```diff
@@ -974,7 +974,11 @@ def push_to_run_queue(self, queue_name, launch_spec):
             for q in queues_found
             if q["name"] == queue_name
             # ensure user has access to queue
-            and (q["access"] == "PROJECT" or q["createdBy"] == self.default_entity)
+            and (
+                # TODO: User created queues in the UI have USER access
+                q["access"] in ["PROJECT", "USER"]
```
[the remark this answers is not in the document]
Comment (hu-po) [reply wandb/sdk/internal/internal_api.py:979]: That is the desired future behavior, but if we remove it now it wont work with user created queues (which is a primary codepath right now).

File: wandb/cli/cli.py [hunk 20]
```diff
@@ -848,6 +860,7 @@ def _parse_settings(settings):
         or env.get("WANDB_ENTITY")
         or config.get("entity")
         or api.settings("entity")
+        or api.default_entity
```
Comment (raubitsj) [inline wandb/cli/cli.py:863]: Slightly risky change... 
The old way, the entity was resolved at the time of the sweep being created, it is actually done in the upsert sweep mutation.

This new way is doing a query from the public api and getting hopefully the same information.... in the future this change is worth noting in the review. 

This is probably fine since this is in the CLI and not the python interface it should be ok as this isnt very optimized and there is no offline mode here.

Comment (hu-po) [reply wandb/cli/cli.py]: Removed.

**Review by raubitsj — APPROVED** [review]
